src/electron/ipc.ts

```diff
@@ -0,0 +1,298 @@
+ *
+ * @example
+ *   // In main process
+ *   const client = new OutlineIpcClient(mainWindow.webContents);
+ *   client.send(APP_UPDATE_DOWNLOADED_CHANNEL);
+ *
+ *   // In renderer process (window.ipcImpl is injected by preload)
+ *   const client = new OutlineIpcClient(window.ipcImpl);
+ *   const result = await client.invoke(STOP_VPN_CHANNEL);
+ */
+export class OutlineIpcClient {
+  /**
+   * Create a new instance of IPC client on top of `ipcImpl`.
+   * @param ipcImpl The underlying implementation to send IPC messages.
+   */
+  public constructor(
+    private readonly ipcImpl: {
+      send(channel: string, ...args: unknown[]): void;
+      invoke?<T extends OutlineIpcTwoWayChannels>(channel: string, ...args: unknown[]): IpcInvokeResultCrossBoundary<T>;
+    }
+  ) {}
+
+  /**
+   * Send a message to the other process via `channel` and return.
+   * @param channel One of the predefined `OUTLINE_ONE_WAY_IPC_CHANNELS`.
+   * @param args Parameters which can be inferred by the `channel`.
+   */
+  public send<T extends OutlineIpcOneWayChannels>(
+    channel: T,
+    ...args: Parameters<OutlineIpcOneWayService[T]>
+  ): ReturnType<OutlineIpcOneWayService[T]> & void {
+    this.ipcImpl.send(channel, ...args);
+  }
+
+  /**
+   * Send a message to main process via `channel` and wait for the result. We
+   * will also convert the returned `ErrorCode` into an `OutlinePluginError`
+   * and throw if it is not `NO_ERROR`.
+   * @param channel One of the predefined `OUTLINE_TWO_WAY_IPC_CHANNELS`.
+   * @param args Parameters which can be inferred by the `channel`.
+   * @returns A promise of the result object returned from the main process.
+   */
+  public async invoke<T extends OutlineIpcTwoWayChannels>(
+    channel: T,
+    ...args: Parameters<OutlineIpcTwoWayService[T]>
+  ): Promise<ReturnType<OutlineIpcTwoWayService[T]>> {
+    const [err, result] = await this.ipcImpl.invoke(channel, ...args);
+    if (err !== ErrorCode.NO_ERROR) {
+      throw new OutlinePluginError(err);
+    }
+    return result as ReturnType<OutlineIpcTwoWayService[T]>;
+  }
+}
+
+/**
+ * The IPC handler class which can be used by both main process and renderer
+ * process. The caller is responsible to give it the correct implementation
+ * of the underlying IPC strategy (for example, `ipcMain`, or an object exposed
+ * by preload script).
+ *
+ * @example
+ *   // In main process
+ *   const handler = new OutlineIpcHandler(ipcMain);
+ *   handler.on(QUIT_APP_CHANNEL, () => { ... });
+ *   handler.handle(STOP_VPN_CHANNEL, async () => { ... });
+ *
+ *   // In renderer process (window.ipcImpl is injected by preload)
+ *   const handler = new OutlineIpcHandler(window.ipcImpl);
+ *   handler.on(APP_UPDATE_DOWNLOADED_CHANNEL, () => { ... });
+ */
+export class OutlineIpcHandler {
+  /**
+   * Create a new instance of IPC handler on top of `ipcImpl`.
+   * @param ipcImpl The underlying implementation to subscribe to IPC messages.
+   */
+  public constructor(
```

Comment:
Add `once`, `on`, `removeListener` methods to the `Platform` interface, so the renderer code can register listeners.

Reply:
No need for an interface on the main process side.